Tests/Algorithm/Framework/Portfolio/MeanReversionPortfolioConstructionModelTest.cs

```diff
@@ -0,0 +1,122 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by aaplicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using NUnit.Framework;
+using Python.Runtime;
+using QuantConnect.Algorithm;
+using QuantConnect.Algorithm.Framework.Alphas;
+using QuantConnect.Algorithm.Framework.Portfolio;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Lean.Engine.DataFeeds;
+using QuantConnect.Lean.Engine.HistoricalData;
+using QuantConnect.Packets;
+using QuantConnect.Tests.Common.Data.UniverseSelection;
+using QuantConnect.Tests.Engine.DataFeeds;
+
+namespace QuantConnect.Tests.Algorithm.Framework.Portfolio
+{
+    [TestFixture]
+    public class MeanReversionPortfolioConstructionModelTests
+    {
+        private DateTime _nowUtc;
+        private QCAlgorithm _algorithm;
+
+        [SetUp]
+        public virtual void SetUp()
+        {
+            _nowUtc = new DateTime(2021, 1, 10);
+            _algorithm = new QCAlgorithm();
+            _algorithm.SetFinishedWarmingUp();
+            _algorithm.SetPandasConverter();
+            _algorithm.SubscriptionManager.SetDataManager(new DataManagerStub(_algorithm));
+            _algorithm.SetDateTime(_nowUtc.ConvertToUtc(_algorithm.TimeZone));
+            _algorithm.SetCash(1200);
+            var historyProvider = new SubscriptionDataReaderHistoryProvider();
+            _algorithm.SetHistoryProvider(historyProvider);
+
+            historyProvider.Initialize(new HistoryProviderInitializeParameters(
+                new BacktestNodePacket(),
+                new Api.Api(),
+                TestGlobals.DataProvider,
+                new SingleEntryDataCacheProvider(new DefaultDataProvider()),
+                TestGlobals.MapFileProvider,
+                TestGlobals.FactorFileProvider,
+                i => { },
+                true,
+                new DataPermissionManager()));
+        }
+
+        [TestCase(Language.CSharp, null, null, 47, 47)]
+        [TestCase(Language.Python, null, null, 47, 47)]
+        [TestCase(Language.CSharp, 1, -0.5, 31, 63)]
+        [TestCase(Language.Python, 1, -0.5, 31, 63)]
+        public void CorrectWeightings(Language language, double? magnitude1, double? magnitude2, decimal expectedQty1, decimal expectedQty2)
```

Comment:
Cool! but think we should add more tests, there are a few cases that would be nice to cover, different arguments `epc,resolution,windowSize`, also see other PCMs test suites 👍

Reply:
I've updated that to include edge cases and keep the theoretical cases suggested by the original author's paper. They should be a good check for the model to work properly. Yet, I dun think toning the hyperparameters and copying the backtested values into it is appropriate.